File: accept.py
import threading
import os
from base64 import b64encode

import requests
import urllib3
import psutil


class Acceptor:

    def __init__(self):
        self.league_proc = 'LeagueClient.exe'
        self.client_username = 'riot'
        self.client_password = ''
        self.procname = ''
        self.pid = -1
        self.host = '127.0.0.1'
        self.port = ''
        self.protocol = ''
        self.headers = ''
        self.session = requests.session()
        self.paused = False
        self.terminate = False
        urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

    def run(self):
        if not self.paused:
            pid = self.get_pid()
            if pid != -1:
                try:
                    if str(pid) != self.pid:
                        self.parse_lockfile()
                    phase = self.request('get', '/lol-gameflow/v1/gameflow-phase').json()
                    if phase == 'ReadyCheck':
                        self.request('post', '/lol-matchmaking/v1/ready-check/accept')
                except:
                    pass

        if not self.terminate:
            threading.Timer(1.0, self.run).start()

    def parse_lockfile(self) -> None:
        lockfile = open(self.find_lockfile_path(), 'r')
        data = lockfile.read()
        lockfile.close()
        data = data.split(':')
        self.procname = data[0]
        self.pid = data[1]
        self.port = data[2]
        self.client_password = data[3]
        self.protocol = data[4]
        userpass = b64encode(bytes('{}:{}'.format(self.client_username, self.client_password), 'utf-8')).decode('ascii')
        self.headers = {'Authorization': 'Basic {}'.format(userpass)}

    def find_lockfile_path(self) -> str:
        path = None
        for pid in psutil.pids():
            if psutil.Process(pid).name() == self.league_proc:
                path = psutil.Process(pid).exe()
        if path is not None:
            return os.path.join(os.path.dirname(path), 'lockfile')

    def request(self, method: str, path: str):
        url = "{}://{}:{}{}".format(self.protocol, self.host, self.port, path)
        fn = getattr(self.session, method)
        r = fn(url, verify=False, headers=self.headers, timeout=3)
        return r

    def get_pid(self):
        for proc in psutil.process_iter():
            if proc.name() == self.league_proc:
                return proc.pid
        return -1

File: src/accept.rs
use std::io::{Error, ErrorKind};
use std::sync::{atomic::{AtomicBool, Ordering}, Arc};
use std::{fs, path::Path, thread, time::Duration};

use base64::encode;
use reqwest::blocking::Client;
use reqwest::header::{HeaderMap, HeaderValue};

#[derive(Clone)]
struct LeagueSession {
    id: u32,
    client: Client,
    addr: String,
}

impl LeagueSession {
    pub fn new(league_pid: u32) -> Result<LeagueSession, Box<dyn std::error::Error>> {
        let mut lockfile_path = String::new();
        unsafe {
            let temp = tasklist::get_proc_path(league_pid);
            let path = Path::new(&temp);
            if let Some(parent) = path.parent() {
                if let Some(parent_str) = parent.to_str() {
                    lockfile_path = format!("{}\\{}", parent_str.to_string(), "lockfile");
                }
            }
        }
        let contents = fs::read_to_string(lockfile_path)?;
        let data: Vec<&str> = contents.split(":").collect();
        if data.len() != 5 {
            return Err(Box::new(Error::new(
                ErrorKind::Other,
                "Could not parse lockfile",
            )));
        }
        let auth = format!(
            "Basic {}",
            encode(format!("{}:{}", "riot".to_string(), data[3].to_string()).as_bytes())
        );
        let url = format!(
            "{}://{}:{}",
            data[4].to_string(),
            String::from("127.0.0.1"),
            data[2].to_string()
        );
        let mut headers = HeaderMap::new();
        headers.insert("Authorization", HeaderValue::from_str(&auth)?);
        let session = Client::builder()
            .default_headers(headers)
            .danger_accept_invalid_certs(true)
            .build()?;
        Ok(LeagueSession {
            id: league_pid,
            client: session,
            addr: url,
        })
    }
}

#[derive(Clone)]
pub struct Acceptor {
    game_pname: String,
    game_pid: u32,
    session: Option<LeagueSession>,
    pub paused: Arc<AtomicBool>,
    pub terminate: Arc<AtomicBool>,
}

impl Acceptor {
    pub fn new() -> Acceptor {
        Acceptor {
            game_pname: String::from("LeagueClient.exe"),
            game_pid: 0,
            session: None,
            paused: Arc::new(AtomicBool::new(false)),
            terminate: Arc::new(AtomicBool::new(false)),
        }
    }

    pub fn run(&mut self) {
        while !self.terminate.load(Ordering::SeqCst) {
            if !self.paused.load(Ordering::SeqCst) && self.game_running() {
                if let Some(_) = self.session.as_ref().filter(|ls| ls.id == self.game_pid) {
                    let phase = self.get_game_phase();
                    if phase == "Matchmaking" {
                        thread::sleep(Duration::from_millis(300));
                        continue;
                    }
                    if phase == "ReadyCheck" {
                        self.accept_match();
                        thread::sleep(Duration::from_secs(1));
                        continue;
                    }
                } else {
                    match LeagueSession::new(self.game_pid) {
                        Ok(session) => {
                            self.session = Some(session);
                        }
                        Err(_) => {}
                    }
                }
            }
            thread::sleep(Duration::from_secs(3));
        }
    }

    fn game_running(&mut self) -> bool {
        unsafe {
            if let Some(pid) = tasklist::find_process_id_by_name(&self.game_pname).get(0) {
                self.game_pid = *pid;
                return true;
            }
        }
        false
    }

    fn get_game_phase(&self) -> String {
        if let Some(session) = &self.session {
            let url = format!("{}{}", session.addr, "/lol-gameflow/v1/gameflow-phase");
            if let Ok(response) = session.client.get(url).send() {
                if let Some(body) = response.text().ok() {
                    return body.trim_matches('"').to_string();
                }
            }
        }
        String::from("Phase not found")
    }

    fn accept_match(&self) {
        if let Some(session) = &self.session {
            let url = format!(
                "{}{}",
                session.addr, "/lol-matchmaking/v1/ready-check/accept"
            );
            let _ = session.client.post(url).send();
        }
    }
}
